diag_manager/fms_diag_buffer.F90

```diff
@@ -0,0 +1,1566 @@
+      allocate(integer(kind=i4_kind) :: this%count_0d(n_samples))
+      this%counter = 0
+      this%typestr = 'i4'
+    type is (integer(kind=i8_kind))
+      allocate(integer(kind=i8_kind) :: this%buffer(buff_sizes(1),buff_sizes(2),buff_sizes(3),buff_sizes(4), &
+                                                  & buff_sizes(5)))
+      allocate(integer(kind=i8_kind) :: this%counter(buff_sizes(1),buff_sizes(2),buff_sizes(3),buff_sizes(4), &
+                                                  & buff_sizes(5)))
+      allocate(integer(kind=i8_kind) :: this%count_0d(n_samples))
+      this%counter = 0
+      this%typestr = 'i8'
+    type is (real(kind=r4_kind))
+      allocate(real(kind=r4_kind) :: this%buffer(buff_sizes(1),buff_sizes(2),buff_sizes(3),buff_sizes(4),  &
+                                               & buff_sizes(5)))
+      allocate(real(kind=r4_kind) :: this%counter(buff_sizes(1),buff_sizes(2),buff_sizes(3),buff_sizes(4), &
+                                                & buff_sizes(5)))
+      allocate(real(kind=r4_kind) :: this%count_0d(n_samples))
+      this%counter = 0
+      this%typestr = 'r4'
+    type is (real(kind=r8_kind))
+      allocate(real(kind=r8_kind) :: this%buffer(buff_sizes(1),buff_sizes(2),buff_sizes(3),buff_sizes(4), &
+                                               & buff_sizes(5)))
+      allocate(real(kind=r8_kind) :: this%counter(buff_sizes(1),buff_sizes(2),buff_sizes(3),buff_sizes(4), &
+                                                & buff_sizes(5)))
+      allocate(real(kind=r8_kind) :: this%count_0d(n_samples))
+      this%counter = 0
+      this%typestr = 'r8'
+    class default
+       call mpp_error("allocate_buffer_5d", &
+           "The buff_type value passed to allocate a buffer is not a r8, r4, i8, or i4",&
+           FATAL)
+  end select
+  allocate(this%num_elements(n_samples))
+  this%num_elements = 0
+  this%count_0d   = 0
+  allocate(this%buffer_dims(5))
+  this%buffer_dims(1) = buff_sizes(1)
+  this%buffer_dims(2) = buff_sizes(2)
+  this%buffer_dims(3) = buff_sizes(3)
+  this%buffer_dims(4) = buff_sizes(4)
+  this%buffer_dims(5) = buff_sizes(5)
+end subroutine allocate_buffer_5d
+
+!! gonna leave these for when we stop caring about gnu <11
+!> @brief Gets buffer data from buffer0d type
+!! @return copy of the buffer data
+!function get_buffer_0d (this) &
+!result(rslt)
+  !class (buffer0d), intent(in) :: this !< scalar buffer object
+  !class(*), allocatable :: rslt
+  !if (allocated(this%buffer)) then
+    !rslt = this%buffer(1)
+  !else
+    !call mpp_error(FATAL, 'get_buffer_0d: buffer not allocated')
+  !endif
+!end function get_buffer_0d
+
+!> Gets real(r4_kind) buffer data from a scalar buffer object
+!! called through buffobj%get_buffer(r4_outputdata)
+subroutine get_0d_real4 (this, buff_out)
+  class(buffer0d), intent(in) :: this !< 0d allocated buffer object
+  real(r4_kind), allocatable, intent(out)  :: buff_out !< output of copied buffer data
```

Comment:
If this was a `class(*)` variable, couldn't you just have a single routine instead of a bunch of them?